### PR TITLE
Bump shellcheck to v0.7.0

### PR DESCRIPTION
**Pull request by odinuge**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Changelog:
https://github.com/koalaman/shellcheck/blob/master/CHANGELOG.md

We have some new erros due to the version bump.

- SC2034: VARIABLE_XYZ appears unused. Verify use (or export if used externally).
    - Applies to all scripts we source in other scripts
- SC2039: In POSIX sh, set option pipefail is undefined.
    - Applies to files using it with "sh" instead of "bash" in the shebang
- SC2054: Use spaces, not commas, to separate array elements.
    - Fixing Should make no difference in the code
- SC2128: Expanding an array without an index only gives the first element.
    - Fixing Should make no difference in the code
- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.
    - Not 100% sure if we can swap to `&& exit 1`. Applies to a lot of
    test code.

All changes should be straight forward to fix, but will be done in a
separate PR.

Errors before updating `hack/.shellcheck_failures`:  https://gist.github.com/odinuge/7e49124448e7dd0cf1d758e2af294a9b

**Which issue(s) this PR fixes**:

Fixes # `N/A`

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
😞

**Comment by BenTheElder** [comment]
> All changes should be straight forward to fix, but will be done in a
separate PR.

if the fixes are trivial, can we just fix them now? when we update eg go we usually fix lint errors then.

**Comment by odinuge** [comment]
> if the fixes are trivial, can we just fix them now? when we update eg go we usually fix lint errors then.

If that is preferred, I can do so. Not 100% sure about  [SC2251](https://github.com/koalaman/shellcheck/wiki/SC2251) tho, so I will take a look

File: hack/.shellcheck_failures
Comment (fejta) [inline hack/.shellcheck_failures]: Why are we adding things? Can we first merge PRs to fix these issues? Ideally we get the whitelist down to zero.

**Review by fejta — COMMENTED** [review]
Thanks!
/approve

File: hack/.shellcheck_failures
Comment (odinuge) [reply hack/.shellcheck_failures]: Yup! Working on it now ref the comment from @BenTheElder ! So ill add the fixes into this PR.

Just doing some investigation on https://github.com/koalaman/shellcheck/wiki/SC2251, that may be masking some test failures. Updating the code as we speak.

We have a bunch of test code like this:

```bash
set -o errexit
[...]
! command_that_should_fail
[...]
```

This test will succeed no matter what the result of `command_that_should_fail` is.

Comment (odinuge) [reply hack/.shellcheck_failures]: Or do you want me to create separate PRs for each fix?

Comment (fejta) [reply hack/.shellcheck_failures]: Either way. In this PR is great too.

File: cluster/gce/gci/mounter/stage-upload.sh
Comment (odinuge) [inline cluster/gce/gci/mounter/stage-upload.sh]: https://github.com/koalaman/shellcheck/wiki/SC2039

We can switch the shebang to bash instead of sh instead of removing this line (I think I am in favor of that). The image used include bash as well.

File: hack/lib/golang.sh
Comment (odinuge) [inline hack/lib/golang.sh]: When shellcheck disable statements are placed here they apply to the whole file (since this applies to several lines)

File: hack/.shellcheck_failures
Comment (odinuge) [reply hack/.shellcheck_failures]: Updated this PR now, but splitted different stuff into different commits. :smile:

File: cluster/gce/gci/mounter/stage-upload.sh
Comment (BenTheElder) [reply cluster/gce/gci/mounter/stage-upload.sh]: +1, let's switch to bash.

File: hack/lib/golang.sh
Comment (BenTheElder) [reply hack/lib/golang.sh]: on the fence about this, we should probably be verifying which we export and which are unused and remove things that are unused.

**Review by fejta — APPROVED** [review]
/lgtm

**Comment by fejta** [comment]
/retest

These failures might be related to this PR

**Comment by odinuge** [comment]
Ref. comment from @BenTheElder about `SC2034`, should I sort that out in this PR or should I create a new PR fixing it?

The PR is only missing approval from `cluster/gce/gci/OWNERS`, but the two people in that owners file look quite inactive. Guess I can move that change into another PR.

/assign dchen1107 yguo0905

**Comment by fejta** [comment]
/assign @bowei @cjcullen mtaufen

for gce approval

**Review by SataQiu — APPROVED** [review]
Great!
/lgtm
/approve

**Comment by BenTheElder** [comment]
yes I think this is OK, I'm not a huge fan of globally disabling lints in files but in this particular case it seems reasonable enough.

thanks for working on this

File: test/images/pets/redis-installer/on-start.sh
Comment (BenTheElder) [inline test/images/pets/redis-installer/on-start.sh]: is this commit identical in behavior?

**Comment by mtaufen** [comment]
/approve
but
/hold
for Ben's latest comment (https://github.com/kubernetes/kubernetes/pull/84249#pullrequestreview-309515954)

File: test/images/pets/redis-installer/on-start.sh
Comment (odinuge) [reply test/images/pets/redis-installer/on-start.sh]: Yes, it is!

Since we use `-a` in read, the result (`LINE`) is an array of the words on the given line.

When `LINE` is an array, `${LINE}` is equivalent to `${LINE[0]}`, and results in the first element in the array.

More info can be found here: https://github.com/koalaman/shellcheck/wiki/SC2128

**Comment by odinuge** [comment]
Ref. https://github.com/kubernetes/kubernetes/pull/84249#discussion_r340991817

/hold cancel

**Pull request merged**
